### ISSUE TITLE
5.3.0 migrations feedback

### ISSUE THREAD

**Issue by Tenzer** [python-social-auth/social-app-django#501]
First off, thank you for maintaining this library!

I thought I would provide a bit of feedback on the migrations included in version 5.3.0, as it was more work to roll out this new version, for us, and required me to make some tweaks due to the size of the `socialauth_usersocialauth` table we have.

I should mention that we are using MySQL 8.0 and the `socialauth_usersocialauth` table contains 46 million rows at the time of writing.

Due to this, it wasn't possible for us to run the migrations as is, since MySQL has to rewrite the data on disk when adding new columns with a default value, and it locks the table for writes during this period. I instead opted for using `pt-online-schema-change` from the Percona Toolkit to first add the new `extra_data_new` column on the table without locking it.

Next up was running the data migration. It required some changes to the code:
```diff
diff --git a/social_django/migrations/0013_migrate_extra_data.py b/social_django/migrations/0013_migrate_extra_data.py
index 2c42171..583e990 100644
--- a/social_django/migrations/0013_migrate_extra_data.py
+++ b/social_django/migrations/0013_migrate_extra_data.py
@@ -9,7 +9,8 @@ def migrate_json_field(apps, schema_editor):
     UserSocialAuth = apps.get_model("social_django", "UserSocialAuth")
     Partial = apps.get_model("social_django", "Partial")
     db_alias = schema_editor.connection.alias
-    for auth in UserSocialAuth.objects.using(db_alias).all():
+    to_be_updated = []
+    for auth in queryset_iterator(UserSocialAuth.objects.using(db_alias).exclude(extra_data='""')):
         old_value = auth.extra_data
         if isinstance(old_value, str):
             try:
@@ -17,7 +18,15 @@ def migrate_json_field(apps, schema_editor):
             except json.JSONDecodeError as error:
                 print(f"Failed to migrate extra_data {old_value}: {error}")
         auth.extra_data_new = old_value
-        auth.save(update_fields=["extra_data_new"])
+        to_be_updated.append(auth)
+
+        if len(to_be_updated) >= 1000:
+            UserSocialAuth.objects.bulk_update(to_be_updated, ["extra_data_new"])
+            to_be_updated.clear()
+
+    if to_be_updated:
+        UserSocialAuth.objects.bulk_update(to_be_updated, ["extra_data_new"])
+
     for auth in Partial.objects.using(db_alias).all():
         old_value = auth.data
         if isinstance(old_value, str):
```

This was to avoid loading all the 46 million rows into memory at once (using `queryset_iterator` from [here](https://djangosnippets.org/snippets/1949/), using `.iterator()` wasn't enough) and to speed up the work by writing data in bulk rather than one row at a time.

When this was done, I could then rename the old and new columns atomically, using a single `ALTER TABLE` statement and roll out the new version of the library.

I'm sure it won't be as painful for other people who are lucky enough to be using PostgreSQL or who don't have that many rows in their tables, but at least there's a chance this may help somebody else who is in a similar situation to us.

I haven't made a pull request with these changes since they may not be relevant for other people, but I thought I would share this regardless.

**Comment by nijel** [comment]
Thanks for the feedback. Using bulk update definitely makes sense. Filtering on `extra_data` is problematic because the old field might have been JSON as well depending on user configuration.

**Comment by megagator** [comment]
@Tenzer Thank you for this amazing write up! I came here to post something detailing a similar story using MySQL 8 and a `usersocialauth` table in the millions. Luckily we weren't storing any data in the extra data column, so the default empty dict was fine, but the migration was still crashing trying to load the entire table at once.

An ideal outcome from this issue (in addition to paginating and bulk updating `0013_migrate_extra_data`) would be to add a testing step somewhere in the release process that runs migrations against a table with millions of rows so that problems like these are caught a lot sooner :)

**Comment by lifenautjoe** [comment]
Hi all, 

We're having the same issue, with some 500k records on our database, the migration breaks our servers. 

Is there any chance these changes can be incorporated in a newer release?

**Comment by nijel** [comment]
If somebody comes up with a pull request, I can turn that into a release.

**Comment by lifenautjoe** [comment]
> When this was done, I could then rename the old and new columns atomically, using a single `ALTER TABLE` statement and roll out the new version of the library.

Hi @Tenzer , could you elaborate on this step? For what I can see, you fixed the memory issue by using an iterator and using a bulk update. 

Why is this needed?

We're putting together a PR and we want to make sure we understand fully your solution.

**Comment by nijel** [comment]
I guess he did single `ALTER TABLE` instead of following two migrations:

* https://github.com/python-social-auth/social-app-django/blob/master/social_django/migrations/0014_remove_usersocialauth_extra_data.py 
* https://github.com/python-social-auth/social-app-django/blob/master/social_django/migrations/0015_rename_extra_data_new_usersocialauth_extra_data.py

**Comment by Tenzer** [comment]
> Hi @Tenzer , could you elaborate on this step? For what I can see, you fixed the memory issue by using an iterator and using a bulk update.

The memory issue was solved by using the iterator instead of loading all rows into memory.

The bulk update helped speed up the process by reducing the number of roundtrips between Django and the database.

I also further sped it up by filtering the existing rows (`.exclude(extra_data='""'))` although @nijel has pointed out that isn't a universally good idea.

> Why is the `ALTER TABLE` needed?

As @nijel mentioned, I replaced the "naive" operations Django would take with more custom steps which worked better for us.

MySQL (in most cases) have to rewrite data on disk when adding or removing columns. That would take a long time in our case and lock the table while it was doing it. So we instead used `pt-online-schema-change` to add the new `extra_data_new` column without locking the table.

When that was done we ran the migration to populate the new columns with the tweaks mentioned in the original post, and we then had to coordinate the deploy which had the updated library version with running:
```sql
ALTER TABLE social_auth_usersocialauth RENAME COLUMN extra_data TO extra_data_old, RENAME COLUMN extra_data_new TO extra_data;
```
This renames the two columns in one step, atomically, and it doesn't require all the table data to be rewritten on disk.

When that was done we could then again use `pt-online-schema-change` to drop the `extra_data_old` column without locking.
